packages/opentelemetry-exporter-prometheus/src/prometheus.ts

```diff
@@ -0,0 +1,280 @@
+        // Currently, _registerMetric creates a new counter every time the value changes,
+        // so the increment here behaves as a set value (increment from 0)
+        metric.inc(this._getLabelValues(labelKeys, ts.labelValues), ts.points[0]
+          .value as number);
+      }
+    }
+
+    if (metric instanceof Prometheus.Gauge) {
+      for (const ts of readableMetric.timeseries) {
+        metric.set(this._getLabelValues(labelKeys, ts.labelValues), ts.points[0]
+          .value as number);
+      }
+    }
+
+    // TODO: only counter and gauge are implemented in metrics so far
+  }
+
+  private _getLabelValues(keys: string[], values: LabelValue[]) {
+    const labelValues: Prometheus.labelValues = {};
+    for (let i = 0; i < keys.length; i++) {
+      if (values[i].value !== null) {
+        labelValues[keys[i]] = values[i].value!;
+      }
+    }
+    return labelValues;
+  }
+
+  private _registerMetric(
+    readableMetric: ReadableMetric
+  ): Prometheus.Metric | undefined {
+    const metricName = this._getPrometheusMetricName(readableMetric.descriptor);
+    const metric = this._registry.getSingleMetric(metricName);
+
+    /**
+     * Prometheus library does aggregation, which means its inc method must be called with
+     * the value to be incremented by. It does not have a set method. As our ReadableMetric
+     * contains the current value, not the value to be incremented by, we destroy and
+     * recreate counters when they are updated.
+     *
+     * This works because counters are identified by their name and no other internal ID
+     * https://prometheus.io/docs/instrumenting/exposition_formats/
+     */
+    if (metric instanceof Prometheus.Counter) {
+      this._registry.removeSingleMetric(metricName);
+    } else if (metric) return metric;
+
+    const newMetric = this._newMetric(readableMetric, metricName);
+    if (!newMetric) return;
+
+    this._registry.registerMetric(newMetric);
+    return newMetric;
+  }
+
+  private _newMetric(
+    readableMetric: ReadableMetric,
+    name: string
+  ): Prometheus.Metric | undefined {
+    const metricObject = {
+      name,
+      help: readableMetric.descriptor.description,
+      labelNames: readableMetric.descriptor.labelKeys,
+    };
+
+    switch (readableMetric.descriptor.type) {
+      case MetricDescriptorType.COUNTER_DOUBLE:
+      case MetricDescriptorType.COUNTER_INT64:
+        return new Prometheus.Counter(metricObject);
+      case MetricDescriptorType.GAUGE_DOUBLE:
+      case MetricDescriptorType.GAUGE_INT64:
+        return new Prometheus.Gauge(metricObject);
+      default:
+        // Other metric types are currently unimplemented
+        return undefined;
+    }
+  }
+
+  private _getPrometheusMetricName(descriptor: MetricDescriptor): string {
+    return this._sanitizePrometheusMetricName(
+      this._prefix ? `${this._prefix}_${descriptor.name}` : descriptor.name
+    );
+  }
+
+  private _sanitizePrometheusMetricName(name: string): string {
```

Comment:
Could you please add a comment about why do you need to sanitize the metric name?

Reply:
Yes I can. It's documented here https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels what the valid names are. More or less, they should match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`. It does say that colons are for use of user defined rules and should not be used by exporters. I might change the validation to reflect these rules and will add a comment.

Reply:
Documented and made it a little stricter/more clear.